test/ETO/ETOCommitment.js

```diff
@@ -4085,31 +4269,62 @@ contract("ETOCommitment", ([, admin, company, nominee, ...investors]) => {
     return divRound(price.mul(Q18.sub(discount)), Q18);
   }
 
-  function calculateMixedTranchePrice(totalAmount, tranche1Amount, tranche1Price, tranche2Price) {
+  function calculateMixedTranchePriceLinear(
+    totalAmount,
+    tranche1Amount,
+    tranche1Price,
+    tranche2Price,
+  ) {
     const tranche2Amount = totalAmount.sub(tranche1Amount);
     // to approximate real mixed price we must also take into account that tokens are quantized so
     // we lose some funds on the edge between two tranches as in each tranche tokens are acquired
     // independently and some funds are lost due to rounding
+    const tokenPower = getTokenPower();
     const tranche1QuantizedAmount = tranche1Amount
-      .div(tranche1Price)
-      .floor()
-      .mul(tranche1Price);
+      .mul(tokenPower)
+      .divToInt(tranche1Price)
+      .mul(tranche1Price)
+      .divToInt(tokenPower);
+    // smart contract tries to recover as much from the first tranche as possible
+    // const recoveredAmount = tranche1Amount.sub(tranche1QuantizedAmount);
     const tranche2QuantizedAmount = tranche2Amount
-      .div(tranche2Price)
-      .floor()
-      .mul(tranche2Price);
+      // .add(recoveredAmount)
+      .mul(tokenPower)
+      .divToInt(tranche2Price)
+      .mul(tranche2Price)
+      .divToInt(tokenPower);
     // still total amount was spent so effective price was higher for a tranche
+    // return non-integer value for precise calculations
     return totalAmount
       .mul(tranche1Price)
       .mul(tranche2Price)
       .div(
         tranche1Price.mul(tranche2QuantizedAmount).add(tranche2Price.mul(tranche1QuantizedAmount)),
-      )
-      .floor();
+      );
+  }
+
+  function calculateMixedTranchePrice(totalAmount, tranche1Amount, tranche1Price, tranche2Price) {
+    // this reproduces how smart contract calculates tokens exactly
+    const tranche2Amount = totalAmount.sub(tranche1Amount);
+    const tokenPower = getTokenPower();
+    const tranche1Tokens = tranche1Amount.mul(tokenPower).divToInt(tranche1Price);
+    const tranche2Tokens = tranche2Amount.mul(tokenPower).divToInt(tranche2Price);
+
+    // console.log(tranche1Tokens.add(tranche2Tokens).toString("10"));
+    const effectivePrice = totalAmount.mul(tokenPower).div(tranche1Tokens.add(tranche2Tokens));
+    // compare to our old linear approx of the same thing
+    expect(
+      calculateMixedTranchePriceLinear(totalAmount, tranche1Amount, tranche1Price, tranche2Price)
+        .sub(effectivePrice)
+        .abs(),
+    ).to.be.bignumber.lt(tokenPower.mul(2));
```

Comment:
I dont understand this. Why is the linear approximation always less then two tokens off?

Reply:
this is also well understood. but it will take a diagram to explain. in ETOTerms contract we compute token amount and round for each tranche separately.

here we compute average weighted price of two tranches and ignore the gap. see `calculateContributionPrivate`

```
// calculate the really spent part to cover for rounding errors in next tranche (remainingAmount)
                    // commented out as with 18 digit scale we have 1 wei difference max and we'll not deploy any other
                    // precisions anymore
                    // discountedAmount = fixedSlotEquityTokenAmount * fixedSlotPrice / EQUITY_TOKEN_POWER;
```

why I even do it? to have very different way of calculating same value. not to just reproduce smart contract algo. to me this is a part of art of test writing.